service/src/main/java/com/google/auto/service/processor/AutoServiceProcessor.java

```diff
@@ -127,26 +131,32 @@ private void processAnnotations(Set<? extends TypeElement> annotations,
     for (Element e : elements) {
       // TODO(gak): check for error trees?
       TypeElement providerImplementer = (TypeElement) e;
-      AnnotationMirror providerAnnotation = getAnnotationMirror(e, AutoService.class).get();
-      DeclaredType providerInterface = getProviderInterface(providerAnnotation);
-      TypeElement providerType = (TypeElement) providerInterface.asElement();
-
-      log("provider interface: " + providerType.getQualifiedName());
-      log("provider implementer: " + providerImplementer.getQualifiedName());
-
-      if (!checkImplementer(providerImplementer, providerType)) {
-        String message = "ServiceProviders must implement their service provider interface. "
-            + providerImplementer.getQualifiedName() + " does not implement "
-            + providerType.getQualifiedName();
-        error(message, e, providerAnnotation);
-      }
+      AnnotationMirror annotationMirror = getAnnotationMirror(e, AutoService.class)
+          .get(); // We already know it's present
+      Set<DeclaredType> providerInterfaces = getFieldOfClasses(
+          annotationMirror,
+          "value",
+          processingEnv.getElementUtils());
+      for (DeclaredType providerInterface : providerInterfaces) {
+        TypeElement providerType = (TypeElement) providerInterface.asElement();
+
+        log("provider interface: " + providerType.getQualifiedName());
+        log("provider implementer: " + providerImplementer.getQualifiedName());
+
+        if (!checkImplementer(providerImplementer, providerType)) {
```

Comment:
What about:

```java
if (checkImplementer(...)) {
  providers.put(...);
} else {
  error(...);
}
```

Reply:
I thought about that, but concerned the behavior change (even though it seems like it's obviously wrong) might be a bit significant if people for some reason relied on it. Done in 7a70c82